presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -294,6 +302,45 @@ private Estimate calculateNullsFractionForPartitioningKey(HiveColumnHandle parti
         return new Estimate(estimatedNullsCount / estimatedTotalRowsCount);
     }
 
+    private Estimate calculateDataSizeForPartitioningKey(HiveColumnHandle partitionColumn, List<HivePartition> queriedPartitions, Estimate rowCount)
+    {
+        String baseType = partitionColumn.getTypeSignature().getBase();
+        if (!VARCHAR.equals(baseType) && !CHAR.equalsIgnoreCase(baseType)) {
+            return Estimate.unknownValue();
+        }
+
+        OptionalDouble averageDataSize = queriedPartitions.stream()
+                .map(HivePartition::getKeys)
+                .map(keys -> keys.get(partitionColumn))
+                .map(NullableValue::getValue)
+                .mapToDouble(object -> object != null ? ((Slice) object).length() : 0)
+                .average();
+
+        if (averageDataSize.isPresent() && !rowCount.isValueUnknown()) {
+            return new Estimate(averageDataSize.getAsDouble() * rowCount.getValue());
+        }
+        return Estimate.unknownValue();
+    }
+
+    private Estimate calculateDataSize(Map<String, PartitionStatistics> statisticsSample, String columnName, Estimate rowCount)
+    {
+        OptionalDouble averageDataSize = statisticsSample.values().stream()
+                .map(PartitionStatistics::getColumnStatistics)
+                .map(columnStatistics -> columnStatistics.get(columnName))
+                .filter(Objects::nonNull)
+                .map(HiveColumnStatistics::getAverageColumnLength)
```

Comment:
> In the statistics collection code i assume that length - is length in bytes. That's the way we interpret this in the optimizer.

In bytes of "memory occupancy". For varchar this means byte count of UTF-8 encoded data.

> In Hive the length is in symbols. And by symbols they don't mean code-points, but the size of char array. Unfortunately there is no reasonable translation can be made. I'm going to just store the number of bytes in the metastore.

This would mean that, for ASCII text, stats computed with Hive will always be 2x what we would calculate. This can be a problem for other tools. Also, this can be problem for ourselves -- not knowing who calculated the stats, we won't know how to interpret them into internal representation (or "just" we will be 2x off).

I am going to ignore this for now, but I think we should revisit this.